test/boost/sstable_compaction_test.cc

```diff
@@ -211,25 +211,30 @@ SEASTAR_TEST_CASE(compact) {
     table_for_tests cf(env.manager(), s);
     auto close_cf = deferred_stop(cf);
 
-        open_sstables(env, s, "test/resource/sstables/compaction", {1,2,3}).then([&env, s, cf, generation] (auto sstables) mutable {
+        auto sstables = open_sstables(env, s, "test/resource/sstables/compaction", {1,2,3}).get();
             auto new_sstable = [&env, gen = make_lw_shared<sstables::generation_type::int_t>(generation), s] {
                 return env.make_sstable(s, (*gen)++);
             };
-            return compact_sstables(sstables::compaction_descriptor(std::move(sstables), default_priority_class()), cf, new_sstable).then([&env, s, cf] (auto) {
+            compact_sstables(sstables::compaction_descriptor(std::move(sstables), default_priority_class()), cf, new_sstable).get();
                 // Verify that the compacted sstable has the right content. We expect to see:
                 //  name  | age | height
                 // -------+-----+--------
                 //  jerry |  40 |    170
                 //    tom |  20 |    180
                 //   john |  20 |   deleted
                 //   nadav - deleted partition
-                return env.reusable_sst(s, generation).then([&env, s] (shared_sstable sst) {
-                    auto reader = make_lw_shared<flat_mutation_reader_v2>(sstable_reader(sst, s, env.make_reader_permit())); // reader holds sst and s alive.
-                    return read_mutation_from_flat_mutation_reader(*reader).then([reader, s] (mutation_opt m) {
+                auto sst = env.reusable_sst(s, generation).get();
+                    auto reader = sstable_reader(sst, s, env.make_reader_permit());
+                    auto close_reader = deferred_close(reader);
+                    auto verify_mutation = [&] (std::function<void(const mutation_opt&)> verify) {
+                        auto m = read_mutation_from_flat_mutation_reader(reader).get();
+                        verify(m);
+                    };
+                    return verify_mutation([&] (const mutation_opt& m) {
                         BOOST_REQUIRE(m);
                         BOOST_REQUIRE(m->key().equal(*s, partition_key::from_singular(*s, data_value(sstring("jerry")))));
                         BOOST_REQUIRE(!m->partition().partition_tombstone());
-                        auto rows = m->partition().clustered_rows();
+                        const auto& rows = m->partition().clustered_rows();
```

Comment:
Previously, the code used a mutation mutation, so `clustered_rows()` resolved to
https://github.com/scylladb/scylladb/blob/1d9b7f3a929bb29454f3f5152785f920f3119fe1/mutation/mutation_partition.hh#L1426
while with `verify_mutation = [&] (std::function<void(const mutation_opt&)>` it resolves to
https://github.com/scylladb/scylladb/blob/1d9b7f3a929bb29454f3f5152785f920f3119fe1/mutation/mutation_partition.hh#L1425

Reply:
I should have broken the patch respectively.
It's crammed with the async thread change, but it is orthogonal indeed.

Reply:
ahh, i see now. thanks!

Reply:
That said, I'm preparing a new version that passes the mutation_opt by value to the verifier function to reduce the churn (and the `verify_mutation` wrapper does nothing further with the mutation_opt so it can just as well pass it on to the verify function)